python/ray/serve/_private/replica_scheduler/pow_2_scheduler.py

```diff
@@ -642,20 +707,43 @@ def maybe_start_scheduling_tasks(self):
         if tasks_to_start > 0:
             self.num_scheduling_tasks_gauge.set(self.curr_num_scheduling_tasks)
 
-    async def choose_replica_for_query(self, query: Query) -> ReplicaWrapper:
+    async def choose_replica_for_request(
+        self, pending_request: PendingRequest, *, is_retry: bool = False
+    ) -> ReplicaWrapper:
         """Chooses a replica to send the provided request to.
 
-        Requests are scheduled in FIFO order, so this puts a future on the internal
-        queue that will be resolved when a replica is available and it's the front of
-        the queue.
+        By default, requests are scheduled in FIFO order, so this places a future on the
+        back of an internal queue that will be popped when a replica is available.
+
+        If `emplace_front` is passed, the request will be placed at the front of the
+        queue.
 
         Upon cancellation (by the caller), the future is cancelled and will be passed
         over when a replica becomes available.
         """
-        pending_request = PendingRequest(asyncio.Future(), query.metadata)
         try:
-            self._pending_requests_to_fulfill.append(pending_request)
-            self._pending_requests_to_schedule.append(pending_request)
+            if not is_retry:
+                self._pending_requests_to_fulfill.append(pending_request)
+                self._pending_requests_to_schedule.append(pending_request)
+            else:
+                index = 0
+                for pr in self._pending_requests_to_fulfill:
```

Comment:
Non-blocker nitpick, we can probably just use `enumerate` and so we don't need to track and increment index separately

Reply:
Also, want to mention, since now that we can assume the queue is sorted. We can utilize python's [bisect.insort](https://docs.python.org/3/library/bisect.html#bisect.insort) 😄

Reply:
Oh nice I hadn't heard of `bisect.insort`. Looks interesting. I think in this case it might not be ideal because in the common case we should be inserting at or near the front of the queue when going through this slower `is_retry` path.